src/runtime_src/core/include/xrt/xrt_kernel.h

```diff
@@ -614,6 +614,43 @@ class run
     start(count);
   }
 
+  /**
+   * read_ctrl_scratchpad() - Read control scratchpad memory
+   * 
+   * @param offset
+   *  Offset in control scratchpad memory
+   * @param size
+   *  Number of bytes to read
+   * @return
+   *  Data read from control scratchpad memory
+   * 
+   * This API is valid only for run objects that are associated with an ELF.
+   * Use this API to read data from control scratchpad memory.
+   * This memory is created by XRT based on ELF used to create xrt::kernel.
+   * 
+   * This function throws if the read fails.
+   */
+  std::vector<char>
```

Comment:
Should we make the type `uint8_t` ?

I am not excited about std::vector as return type.   Reason is that doesn't handle a case where application controls the data storage, say a user pointer.  Using a std::vector could lead to unnecessary memcpy.

Thinking out loud here, I wonder if the exchange should be through an xrt::bo?  Obviously not a BO that is created by driver, but a type of BO that holds just system memory, either allocated when the BO is created or as a userptr passed by application?

Reply:
I don't think a BO makes sense.  The data is not managed by a BO.  Once written the connection to the BO is lost, I think.

Reply:
Yes Soren, I agree this approach is simple and let user manage bo however he wants. Thanks for the suggestion.
I will rewrite the code accordingly.